### ISSUE TITLE
v16: target_effective_lag cert penalty missing — cert.certified_liq_deficit=0 blocks liquidation on AuthMark assets (deployed BPF a12faf78 lacks your own 323c9f2)

### ISSUE THREAD

**Issue by LGopus** [aeyakovenko/percolator#68]
> **Edit log:** corrected (1) line numbers for the Vec-runtime sites, zero-copy `recertify_account_after_trade_delta`, AND the `risk_notional_ceil` reference — earlier I cited my patched working copy; correct values for unpatched `9bcf002b` are inline below; (2) lag-decay slot count (~44 slots geometric, not ~417 — math error); (3) removed an unsourced framing quote; (4) replaced the body code block with exact upstream source text (was a compressed `?)?` chain); (5) corrected helper return types from `Result<u128>` → `V16Result<u128>` (the former isn't valid Rust syntax); (6) corrected sequence count "three" → "four" instructions (UpdateAssetLifecycle + ConfigureAuthMark + PushAuthMark + PermissionlessCrank); (7) clarified that the PoC's lag arises after a 1-slot effective_price clamp-walk; (8) corrected wrapper deployed rev `0925ed4` → `6512fa1`; (9) corrected the 323c9f2 status from "unmerged" → "merged to `master` HEAD but not deployed" (it IS the current HEAD of the default branch). SHAs, the lag-penalty math, and three-way regression-guard pair-verification are unaffected.

## Summary

`compute_account_health_cert_with_price_override` (engine `v16.rs:6953`), `refresh_account_and_certify_not_atomic` (engine `v16.rs:7033`), `recertify_account_after_trade_delta` (zero-copy ViewMut, engine `v16.rs:9688`) — plus the Vec-runtime twins (`v16.rs:12690, 13789`, and the cert site at `v16.rs:14418` inside `settle_account_for_position_action_and_refresh_not_atomic`) — compute margin requirements from `effective_price` only. They do NOT consult `raw_oracle_target_price`. So when the per-slot price clamp (`max_price_move_bps_per_slot`) holds `effective_price` away from the un-clamped target, the realized loss in transit is **omitted from the cert**.

```rust
// engine v16.rs:6983-7000 (pre-fix, deployed) — exact source text
let risk_notional = risk_notional_ceil(leg.basis_pos_q.unsigned_abs(), price)?;
initial_req = initial_req
    .checked_add(margin_requirement(
        risk_notional,
        config.initial_margin_bps,
        config.min_nonzero_im_req,
    )?)
    .ok_or(V16Error::ArithmeticOverflow)?;
maintenance_req = maintenance_req
    .checked_add(margin_requirement(
        risk_notional,
        config.maintenance_margin_bps,
        config.min_nonzero_mm_req,
    )?)
    .ok_or(V16Error::ArithmeticOverflow)?;
worst_case_loss = worst_case_loss
    .checked_add(risk_notional)
    .ok_or(V16Error::ArithmeticOverflow)?;
// raw_oracle_target_price is NEVER consulted in this block
```

Consequence: an AuthMark-asset position with adverse lag — e.g. a long whose oracle target dropped 10% below `effective_price` — reports `certified_liq_deficit == 0` even when the account is materially underwater. Since liquidation is gated by `if cert.certified_liq_deficit == 0 { return Err(V16Error::NonProgress) }` (engine `v16.rs:9467`), **liquidation is blocked** for the entire mark-walk duration.

The fix exists in your own `323c9f2 "Harden v16 engine invariants"` (Anatoly, 2026-05-26T21:44:26Z) — currently the HEAD of `aeyakovenko/percolator @ master` — but is **NOT in the deployed BPF**. The deployed binary at `a12faf78…` already includes `6512fa1` (the now-slot authentication fix, wrapper-side, 2026-05-26 18:55Z), so a redeploy has happened, but without `323c9f2`.

## Target

| | |
|---|---|
| Mainnet program | `4m3ipBQDYX6JQ9YSmUXDjESDHMtGWtiXforkWr9Qoxdi` |
| Deployed BPF SHA-256 | `a12faf7867c67feecf079391e6a9665efdbe8722562c9b6abdc51a0d6a798882` |
| Wrapper deployed rev | `aeyakovenko/percolator-prog @ 6512fa1` (= `0925ed4` + the now-slot authentication patch; `0925ed4` is the immediate parent) |
| Engine pinned | `aeyakovenko/percolator @ 9bcf002b` (SHA-256 of `src/v16.rs`: `8d5559e2caf2006f92b2eef2d37bf6a30e90c50710854bd140d6fafe8028f997`) |
| Your fix (merged to `master` HEAD, **not deployed**) | `aeyakovenko/percolator @ 323c9f2` |
| Verified | `solana program dump … --url mainnet-beta` at 2026-05-27T02:50Z |

## Reachability

Four-instruction sequence (which can fit in a single transaction) creates the lag state on any permissionless-activated AuthMark asset:

1. `UpdateAssetLifecycle(ACTIVATE)` — pays `init_fee`, attacker holds `oracle_authority`
2. `ConfigureAuthMark(asset_index, initial_mark_e6)`
3. `PushAuthMark(asset_index, mark_e6: <adverse>)` + `PermissionlessCrank(Refresh)`

The wrapper's `hybrid_effective_price_for_crank_view` (`v16_program.rs:10098`) computes a clamped `effective_price` (per `max_price_move_bps_per_slot`) and writes `asset.raw_oracle_target_price = un-clamped target` (`v16_program.rs:9196`). `raw ≠ effective` until effective walks to target; persistence in slots scales with the clamp value.

**Exploitability vs. clamp:** PushAuthMark imposes no clamp on `raw_oracle_target_price`, so any lag magnitude can be created in a single transaction. The clamp `max_price_move_bps_per_slot` governs only how fast `effective_price` follows. Lag persistence (slots until effective catches up) is approximately `ln(target_ratio) / ln(1 - clamp/10000)`. Examples for a 10% target move:

| clamp_bps | slots to close | ~wall-clock (400ms/slot) |
|-----------|----------------|---------------------------|
| 24        | ~44            | ~17.5 s                   |
| 100       | ~10            | ~4.0 s                    |
| 1         | ~1054          | ~7.0 min                  |
| 10000     | 1              | ~0.4 s                    |

The PoC harness uses `max_price_move_bps_per_slot = 24` (per `Bounty5Env::new` in `tests/bounty6_pocs.rs`). The deployed mainnet market's actual value sets the real-world persistence window.

After the PoC sequence: `asset[1].effective_price = 99,760,000`, `asset[1].raw_oracle_target_price = 90,000,000`, lag = 9,760,000 atoms (adverse for long, after one slot of effective_price clamp-walk following PushAuthMark + Crank in the same slot).

## Single-variable causality proof

`bounty6_lag_penalty_extensive_matrix` (4 scenarios, all other gating flags asserted unchanged):

| Row | Side  | Lag       | cert.mm_req on deployed | Post-fix mm_req | Match |
|-----|-------|-----------|--------------------------|------------------|-------|
| 1   | LONG  | 0         | 5,000,000                | 5,000,000        | identical (lag=0) |
| 2   | LONG  | 9,760,000 (raw<eff) | **4,988,000**  | **14,748,000**   | **pre-fix on deployed** |
| 3   | SHORT | 0         | 5,000,000                | 5,000,000        | identical (lag=0) |
| 4   | SHORT | 9,760,000 (raw>eff) | **5,012,000**  | **14,772,000**   | **pre-fix on deployed** |

Math: post-fix mm_req = `5% × effective_price + |adverse_delta| × abs_pos_q / POS_SCALE` (with 1-unit position, the lag penalty equals the lag value).

The `cert.certified_liq_deficit` reads as `0` (deployed) vs `8,758,000` (post-fix) on Row 2 — the latter is the actual deficit the system would correctly liquidate.

## Three-way verification

| BPF | SHA-256 (prefix) | regression-guard test | mm_req |
|-----|------------------|------------------------|--------|
| Deployed mainnet | `a12faf78…` | **FAIL** | 4,988,000 |
| Local build of `0925ed4 + 9bcf002b` (wrapper parent of the deployed rev) | `1bc4caaf…` | **FAIL** | 4,988,000 |
| Local build with minimal port of `323c9f2` | `2393292e…` | **PASS** | 14,748,000 |

Test: `bounty6_lag_penalty_fix_regression_guard` in the LiteSVM harness.

## Why this is a (non-insurance) drain primitive

Attacker as oracle authority on a permissionless-activated AuthMark market:

1. Opens position against a counterparty (or honest LP).
2. Pushes `AuthMark` target adversely against THEIR own position direction — creating arbitrary instantaneous lag.
3. Counterparty's gain crystallizes at the clamped `effective_price`, NOT at the un-clamped `raw_oracle_target_price`.
4. Attacker walks away with realized loss capped at how far `effective_price` has walked, not the full target-implied loss.

**Counterparty fund loss** — under-paid by the lag amount at the moment of close. Magnitude per close = `lag × abs_pos_q / POS_SCALE`. Asymmetric in attacker's favor when attacker controls the AuthMark.

**NOT a strict insurance drain** — per-domain insurance isolation (`8e0e3f8 "Isolate v16 insurance domain budgets"`) bounds liquidation pay by the asset's own domain budget. New asset domains start at budget=0 (init_fee credits `domain 0` only — wrapper `v16_program.rs:7749, 7784`). Empirical: full chain via `bounty6_lag_bypass_chain_drain_attempt` ended with `insurance: 0 → 12,100` (attacker contributed 20,100 in init_fee + trade fees, liquidation paid 8,000 from asset's domain budget). Net positive.

Per the bounty win condition (*"Cause `engine.insurance_fund.balance` to drop below its current value via any sequence of public-instruction calls. Insurance fund balance drop. NOT user fund loss. NOT user fund lockup."*), this is **out of scope** — it is user-fund-loss class via AuthMark isolation break. **Filing as defensive disclosure; eligibility your call** (could conceivably fall under your in-scope "Mark-EWMA exploitation" category, but I can't push it to a strict insurance drain).

## Fix

Minimal port of `323c9f2` (your own) against engine `9bcf002b`:

- 267-line diff, 8 hunks, applies cleanly with `patch -p1`
- **6 cert sites patched** (3 zero-copy compiled into BPF + 3 Vec-runtime for Kani parity)
- **3 new helper functions** (free fns inserted near `risk_notional_ceil` which is at `v16.rs:19734` in `9bcf002b`):
  - `target_effective_lag_adverse_delta(side, effective, raw_target) -> u64`
  - `target_effective_lag_loss_penalty(abs_pos_q, side, effective, raw_target) -> V16Result<u128>`
  - `health_requirements_from_notional_and_target_lag(config, notional, lag_penalty) -> V16Result<(u128, u128, u128)>`
- Centralized so the penalty cannot be accidentally omitted at any cert site

Available on request — happy to paste inline or open a PR.

## Reproduction

```bash
# Build wrapper against engine 9bcf002b
# (Use any wrapper rev whose Cargo.toml pins the engine to 9bcf002b — e.g. 0925ed4 or 6512fa1)
cd /path/to/percolator-prog && cargo build-sbf --no-default-features

# Fetch deployed binary
solana program dump 4m3ipBQDYX6JQ9YSmUXDjESDHMtGWtiXforkWr9Qoxdi /tmp/deployed.so --url mainnet-beta
cp /tmp/deployed.so target/deploy/percolator_prog.so

# Run the regression-guard against deployed BPF — fails (bug live)
cargo test --test bounty6_pocs --release bounty6_lag_penalty_fix_regression_guard -- --nocapture
# Expected: "FIX MISSING: mm_req=4988000 but expected 14748000 (lag penalty omitted)"

# Run the matrix — confirms bug on both long+adverse and short+adverse, single-variable causality
cargo test --test bounty6_pocs --release bounty6_lag_penalty_extensive_matrix -- --nocapture
```

**Comment by aeyakovenko** [comment]
Closing this as already fixed in the current engine source.

Clean-room local verification on `master` shows the reported target/effective-lag liquidation-cert gap is covered by the existing `323c9f2` engine changes:

- Current `src/v16.rs` applies `target_effective_lag_loss_penalty` in runtime and zero-copy health certification paths.
- `cargo test --features test --test v16_spec_tests target_effective_lag -- --nocapture` passes: 6/6 target/effective-lag tests.
- `cargo kani --tests --features fuzz --exact --harness proof_v16_target_effective_lag_adverse_loss_enters_leg_health_requirement --output-format terse` passes with 2/2 cover properties satisfied.

So this is a valid historical/deployment-lag report against older pinned/deployed code, but not an open engine implementation bug in this repository. Any remaining action is deployment/versioning outside this engine issue.
